File: ports/ruby/lib/unicode_ruby/security/display/filename_disguise.rb
# frozen_string_literal: true

require_relative "../covert/bidi_control_balance"
require_relative "../../segmentation/grapheme"

module UnicodeRuby
  module Security
    module Display
      # FilenameDisguise — detection of filename/extension disguise attacks
      # where the visible extension differs from the byte extension (the
      # display-layer detector).
      #
      # Byte-faithful port of the verified Rust reference implementation and of
      # its Lean source of truth.
      #
      # Threat model.  An adversary delivers a file whose rendered name looks
      # like a benign type (`document.txt`) but whose actual byte extension is
      # executable — the canonical attack inserts U+202E RIGHT-TO-LEFT OVERRIDE
      # so `document<RLO>txt.exe` renders as `document exe.txt`.
      #
      # Detection is presentation- and language-agnostic: it surfaces every
      # codepoint that could cause display-vs-byte divergence in the filename —
      # any bidi format-control anywhere, and any fullwidth/halfwidth or
      # combining (grapheme Extend) codepoint in the extension region (after the
      # last `.`).  Native-RTL names with no bidi controls clear.  It reuses the
      # port's own predicates — the bidi format-control set
      # (`Covert::BidiControlBalance`), the grapheme Extend class
      # (`Segmentation::Grapheme`), and the inline fullwidth range — never a host
      # filesystem or rendering library.
      #
      # Sub-threats (priority order):
      #   1. RloFlip            any bidi format-control in the input.
      #   2. WidthClassExt      a fullwidth/halfwidth codepoint in the extension.
      #   3. CombiningInExt     a combining (Extend) codepoint in the extension.
      #   4. MultipleExtensions >= 3 dots (advisory; e.g. `.tar.gz.sig`).
      module FilenameDisguise
        # The three-or-more-dot bound at which the advisory MultipleExtensions
        # sub-threat fires.
        MIN_EXTENSION_DOTS = 3

        # The sub-threat kind discriminants, in priority order.  Each maps to a
        # stable fixture-row / wire tag through `SubThreat#tag`.
        module SubKind
          RLO_FLIP = :rlo_flip
          WIDTH_CLASS_EXT = :width_class_ext
          COMBINING_IN_EXT = :combining_in_ext
          MULTIPLE_EXTENSIONS = :multiple_extensions
        end

        # A sub-threat this detector can fire.  `kind` is the discriminant;
        # `data` carries the variant-specific payload (`position` + `control_cp`
        # for RloFlip, `position` + `cp` for WidthClassExt / CombiningInExt,
        # `dot_count` for MultipleExtensions).  `tag` maps the discriminant to
        # the stable wire tag with an exhaustive switch.
        SubThreat = Struct.new(:kind, :data) do
          def tag
            case kind
            when SubKind::RLO_FLIP then "RloFlip"
            when SubKind::WIDTH_CLASS_EXT then "WidthClassExt"
            when SubKind::COMBINING_IN_EXT then "CombiningInExt"
            when SubKind::MULTIPLE_EXTENSIONS then "MultipleExtensions"
            else
              raise "FilenameDisguise::SubThreat#tag: unknown kind #{kind.inspect}"
            end
          end
        end

        # Top-level classification.  `sub` is nil when clear; `positions` is the
        # codepoint indices the sub-threat implicates (empty when clear);
        # `decoded` is the decoded-byte projection (always empty for this
        # detector, kept for shape parity with the Lean `Classification.hazard`).
        Classification = Struct.new(:sub, :positions, :decoded) do
          def clear?
            sub.nil?
          end

          def tag
            sub.nil? ? nil : sub.tag
          end
        end

        # The structured output of `detect` (mirrors the Lean `Verdict`).
        Verdict = Struct.new(
          :input, :classify, :dot_positions, :last_dot_pos,
          :bidi_control_count, :fullwidth_in_ext, :combining_in_ext
        )

        module_function

        # ── Sub-threat constructors ────────────────────────────────────────

        # A bidi format-control at `position` (codepoint `control_cp`).
        def rlo_flip(position, control_cp)
          SubThreat.new(SubKind::RLO_FLIP, { position: position, control_cp: control_cp })
        end

        # A fullwidth/halfwidth codepoint in the extension at `position`
        # (codepoint `cp`).
        def width_class_ext(position, cp)
          SubThreat.new(SubKind::WIDTH_CLASS_EXT, { position: position, cp: cp })
        end

        # A combining (grapheme Extend) codepoint in the extension at `position`
        # (codepoint `cp`).
        def combining_in_ext(position, cp)
          SubThreat.new(SubKind::COMBINING_IN_EXT, { position: position, cp: cp })
        end

        # Three or more `.` separators (advisory).
        def multiple_extensions(dot_count)
          SubThreat.new(SubKind::MULTIPLE_EXTENSIONS, { dot_count: dot_count })
        end

        # ── Classification constructors ────────────────────────────────────

        def clear
          Classification.new(nil, [], [])
        end

        def hazard(sub, positions, decoded)
          Classification.new(sub, positions, decoded)
        end

        # ── Core predicates (reuse the port's own tables) ──────────────────

        # True iff `cp` is U+002E FULL STOP (the extension separator).
        def ascii_dot?(cp)
          cp == 0x002E
        end

        # True iff `cp` is in the Halfwidth and Fullwidth Forms block.
        def fullwidth_halfwidth?(cp)
          cp >= 0xFF01 && cp <= 0xFFEF
        end

        # True iff `cp` is a bidi format-control (reuses the port's own
        # LRE/RLE/LRO/RLO/PDF/LRI/RLI/FSI/PDI predicate).
        def bidi_format_control?(cp)
          Covert::BidiControlBalance.bidi_format_control?(cp)
        end

        # True iff `cp` has Grapheme_Cluster_Break = Extend (reuses the port's
        # UAX #29 grapheme table).
        def grapheme_extend?(cp)
          Segmentation::Grapheme.lookup_gcb(cp) == :extend
        end

        # ── Sub-detectors ──────────────────────────────────────────────────

        # Positions of every `.` in `input`.
        def dot_positions(input)
          out = []
          input.each_index { |idx| out << idx if ascii_dot?(input[idx]) }
          out
        end

        # Position and codepoint of the first bidi format-control, or nil.
        def first_bidi_control(input)
          input.each_index do |idx|
            return [idx, input[idx]] if bidi_format_control?(input[idx])
          end
          nil
        end

        # Position and codepoint of the first fullwidth/halfwidth codepoint at or
        # after `start`, or nil.
        def first_fullwidth_from(input, start)
          input.each_index do |idx|
            return [idx, input[idx]] if idx >= start && fullwidth_halfwidth?(input[idx])
          end
          nil
        end

        # Position and codepoint of the first Extend codepoint at or after
        # `start`, or nil.
        def first_extend_from(input, start)
          input.each_index do |idx|
            return [idx, input[idx]] if idx >= start && grapheme_extend?(input[idx])
          end
          nil
        end

        # Count of fullwidth/halfwidth codepoints at or after `start`.
        def count_fullwidth_from(input, start)
          count = 0
          input.each_index { |idx| count += 1 if idx >= start && fullwidth_halfwidth?(input[idx]) }
          count
        end

        # Count of Extend codepoints at or after `start`.
        def count_extend_from(input, start)
          count = 0
          input.each_index { |idx| count += 1 if idx >= start && grapheme_extend?(input[idx]) }
          count
        end

        # ── Top-level detection ────────────────────────────────────────────

        # The FilenameDisguise detection function.
        def detect(input)
          dots = dot_positions(input)
          last_dot = dots.last
          ext_start = last_dot.nil? ? input.length : last_dot + 1
          bidi_count = input.count { |cp| bidi_format_control?(cp) }
          fw_in_ext = count_fullwidth_from(input, ext_start)
          ext_in_ext = count_extend_from(input, ext_start)

          Verdict.new(
            input.dup, classify(input, dots, ext_start), dots, last_dot,
            bidi_count, fw_in_ext, ext_in_ext
          )
        end

        # The priority ladder, factored out of `detect`.
        def classify(input, dots, ext_start)
          # Priority 1: any bidi format-control.
          bidi = first_bidi_control(input)
          unless bidi.nil?
            pos, ctl_cp = bidi
            return hazard(rlo_flip(pos, ctl_cp), [pos], [])
          end

          # Priority 2: fullwidth/halfwidth in the extension.
          fw = first_fullwidth_from(input, ext_start)
          unless fw.nil?
            pos, cp = fw
            return hazard(width_class_ext(pos, cp), [pos], [])
          end

          # Priority 3: combining mark in the extension.
          ext = first_extend_from(input, ext_start)
          unless ext.nil?
            pos, cp = ext
            return hazard(combining_in_ext(pos, cp), [pos], [])
          end

          # Priority 4: three or more extensions (advisory).
          if dots.length >= MIN_EXTENSION_DOTS
            return hazard(multiple_extensions(dots.length), dots.dup, [])
          end

          clear
        end
      end
    end
  end
end

File: ports/python/src/unicode_python/security/display/filename_disguise.py
"""FilenameDisguise — detection of filename/extension disguise attacks
where the visible extension differs from the byte extension (the
display-layer detector).

Byte-faithful transliteration of the verified Rust reference
implementation (itself a transliteration of
``Unicode.Security.Display.FilenameDisguise``).

Threat model. An adversary delivers a file whose rendered name looks like
a benign type (``document.txt``) but whose actual byte extension is
executable — the canonical attack inserts ``U+202E`` RIGHT-TO-LEFT
OVERRIDE so ``document<RLO>txt.exe`` renders as ``document exe.txt``.

Detection is presentation- and language-agnostic: it surfaces every
codepoint that could cause display-vs-byte divergence in the filename —
any bidi format-control anywhere, and any fullwidth/halfwidth or
combining (grapheme Extend) codepoint in the extension region (after the
last ``.``). Native-RTL names with no bidi controls clear. It reuses the
port's own predicates (the bidi-format-control set, the grapheme Extend
class, the fullwidth range), never a host filesystem or rendering library.

Sub-threats (priority order):
    1. ``RloFlip``            any bidi format-control in the input.
    2. ``WidthClassExt``      a fullwidth/halfwidth codepoint in the extension.
    3. ``CombiningInExt``     a combining (Extend) codepoint in the extension.
    4. ``MultipleExtensions`` >= 3 dots (advisory; e.g. ``.tar.gz.sig``).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Union

from ..calculus import ClassificationKind
from ..covert import bidi_control_balance
from ...segmentation import grapheme

# ─────────────────────────────────────────────────────────────────────
# §1 Constants
# ─────────────────────────────────────────────────────────────────────

# The number of ``.`` separators at or beyond which a name is treated as
# an advisory MultipleExtensions hazard.
MIN_MULTIPLE_EXTENSIONS = 3


# ─────────────────────────────────────────────────────────────────────
# §2 Sub-threat ADT + verdict
# ─────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class RloFlip:
    """A bidi format-control at ``position`` (codepoint ``control_cp``)."""

    position: int
    control_cp: int


@dataclass(frozen=True, slots=True)
class WidthClassExt:
    """A fullwidth/halfwidth codepoint in the extension, at ``position``
    (codepoint ``cp``)."""

    position: int
    cp: int


@dataclass(frozen=True, slots=True)
class CombiningInExt:
    """A combining (grapheme Extend) codepoint in the extension, at
    ``position`` (codepoint ``cp``)."""

    position: int
    cp: int


@dataclass(frozen=True, slots=True)
class MultipleExtensions:
    """Three or more ``.`` separators (advisory); ``dot_count`` is the
    number of separators."""

    dot_count: int


SubThreat = Union[
    RloFlip,
    WidthClassExt,
    CombiningInExt,
    MultipleExtensions,
]


def sub_threat_tag(sub: SubThreat) -> str:
    """Fixture-row tag string for a sub-threat (matches ``SubThreat.tag``
    in the Lean/Rust reference)."""
    if isinstance(sub, RloFlip):
        return "RloFlip"
    if isinstance(sub, WidthClassExt):
        return "WidthClassExt"
    if isinstance(sub, CombiningInExt):
        return "CombiningInExt"
    if isinstance(sub, MultipleExtensions):
        return "MultipleExtensions"
    raise TypeError(f"sub_threat_tag: unknown SubThreat variant {sub!r}")


@dataclass(frozen=True, slots=True)
class Classification:
    """Top-level FilenameDisguise classification. ``is_clear``
    distinguishes the ``Clear`` variant from the ``Hazard`` variant; a
    hazard carries its sub-threat, the codepoint positions it implicates,
    and the decoded-byte projection (always empty here, kept for shape
    parity with the Lean ``Classification.hazard``)."""

    is_clear: bool
    sub: SubThreat | None = None
    positions: list[int] = field(default_factory=list)
    decoded: list[int] = field(default_factory=list)

    @property
    def tag(self) -> str | None:
        """Human-facing tag for a hazard, or ``None`` when clear."""
        if self.is_clear or self.sub is None:
            return None
        return sub_threat_tag(self.sub)

    @property
    def kind(self) -> ClassificationKind:
        """The classification kind (``CLEAR`` or ``HAZARD``)."""
        return (
            ClassificationKind.CLEAR
            if self.is_clear
            else ClassificationKind.HAZARD
        )


@dataclass(frozen=True, slots=True)
class Verdict:
    """The structured output of :func:`detect` (mirrors the Lean
    ``Verdict``)."""

    input: list[int]
    classify: Classification
    dot_positions: list[int]
    last_dot_pos: int | None
    bidi_control_count: int
    fullwidth_in_ext: int
    combining_in_ext: int


# ─────────────────────────────────────────────────────────────────────
# §3 Core predicates
# ─────────────────────────────────────────────────────────────────────


def is_ascii_dot(cp: int) -> bool:
    """True iff ``cp`` is ``U+002E FULL STOP`` (the extension separator)."""
    return cp == 0x002E


def is_fullwidth_halfwidth(cp: int) -> bool:
    """True iff ``cp`` is in the Halfwidth and Fullwidth Forms block."""
    return 0xFF01 <= cp <= 0xFFEF


def is_bidi_format_control(cp: int) -> bool:
    """True iff ``cp`` is a bidi format-control (reuses the port's own
    :func:`bidi_control_balance.is_bidi_format_control`)."""
    return bidi_control_balance.is_bidi_format_control(cp)


def is_grapheme_extend(cp: int) -> bool:
    """True iff ``cp`` has ``Grapheme_Cluster_Break = Extend`` (reuses the
    port's grapheme table via :func:`grapheme.is_grapheme_extend`)."""
    return grapheme.is_grapheme_extend(cp)


# ─────────────────────────────────────────────────────────────────────
# §4 Sub-detectors
# ─────────────────────────────────────────────────────────────────────


def _dot_positions(input_cps: list[int]) -> list[int]:
    """Positions of every ``.`` in ``input_cps``."""
    return [index for index, cp in enumerate(input_cps) if is_ascii_dot(cp)]


def _first_bidi_control(input_cps: list[int]) -> tuple[int, int] | None:
    """Position and codepoint of the first bidi format-control."""
    for index, cp in enumerate(input_cps):
        if is_bidi_format_control(cp):
            return (index, cp)
    return None


def _first_fullwidth_from(
    input_cps: list[int], start: int
) -> tuple[int, int] | None:
    """Position and codepoint of the first fullwidth/halfwidth codepoint
    at or after ``start``."""
    for index, cp in enumerate(input_cps):
        if index >= start and is_fullwidth_halfwidth(cp):
            return (index, cp)
    return None


def _first_extend_from(
    input_cps: list[int], start: int
) -> tuple[int, int] | None:
    """Position and codepoint of the first Extend codepoint at or after
    ``start``."""
    for index, cp in enumerate(input_cps):
        if index >= start and is_grapheme_extend(cp):
            return (index, cp)
    return None


def _count_fullwidth_from(input_cps: list[int], start: int) -> int:
    """Count of fullwidth/halfwidth codepoints at or after ``start``."""
    return sum(
        1
        for index, cp in enumerate(input_cps)
        if index >= start and is_fullwidth_halfwidth(cp)
    )


def _count_extend_from(input_cps: list[int], start: int) -> int:
    """Count of Extend codepoints at or after ``start``."""
    return sum(
        1
        for index, cp in enumerate(input_cps)
        if index >= start and is_grapheme_extend(cp)
    )


# ─────────────────────────────────────────────────────────────────────
# §5 Top-level detection
# ─────────────────────────────────────────────────────────────────────


def detect(input_cps: list[int]) -> Verdict:
    """The FilenameDisguise detection function."""
    dots = _dot_positions(input_cps)
    last_dot = dots[-1] if dots else None
    ext_start = last_dot + 1 if last_dot is not None else len(input_cps)

    bidi_count = sum(1 for cp in input_cps if is_bidi_format_control(cp))
    fw_in_ext = _count_fullwidth_from(input_cps, ext_start)
    ext_in_ext = _count_extend_from(input_cps, ext_start)

    bidi_hit = _first_bidi_control(input_cps)
    if bidi_hit is not None:
        # Priority 1: any bidi format-control.
        pos, ctl_cp = bidi_hit
        classification = Classification(
            is_clear=False,
            sub=RloFlip(position=pos, control_cp=ctl_cp),
            positions=[pos],
        )
    else:
        fw_hit = _first_fullwidth_from(input_cps, ext_start)
        if fw_hit is not None:
            # Priority 2: fullwidth/halfwidth in the extension.
            pos, cp = fw_hit
            classification = Classification(
                is_clear=False,
                sub=WidthClassExt(position=pos, cp=cp),
                positions=[pos],
            )
        else:
            ext_hit = _first_extend_from(input_cps, ext_start)
            if ext_hit is not None:
                # Priority 3: combining mark in the extension.
                pos, cp = ext_hit
                classification = Classification(
                    is_clear=False,
                    sub=CombiningInExt(position=pos, cp=cp),
                    positions=[pos],
                )
            elif len(dots) >= MIN_MULTIPLE_EXTENSIONS:
                # Priority 4: three or more extensions (advisory).
                classification = Classification(
                    is_clear=False,
                    sub=MultipleExtensions(dot_count=len(dots)),
                    positions=list(dots),
                )
            else:
                classification = Classification(is_clear=True)

    return Verdict(
        input=list(input_cps),
        classify=classification,
        dot_positions=dots,
        last_dot_pos=last_dot,
        bidi_control_count=bidi_count,
        fullwidth_in_ext=fw_in_ext,
        combining_in_ext=ext_in_ext,
    )
